tsl/src/continuous_aggs/materialize.c

```diff
@@ -330,404 +418,386 @@ build_merge_update_clause(List *column_names)
-								2,
-								types,
-								values,
-								nulls,
-								false /* read_only */,
-								0 /* count */);
+static void
+emit_materialization_insert_error(MaterializationContext *context)
+{
+	elog(ERROR,
+		 "could not insert old values into materialization table \"%s.%s\"",
+		 NameStr(*context->materialization_table.schema),
+		 NameStr(*context->materialization_table.name));
+}
 
-	if (res < 0)
-		elog(ERROR,
-			 "could not delete values from the materialization table \"%s.%s\"",
-			 NameStr(*materialization_table.schema),
-			 NameStr(*materialization_table.name));
-	else
-		elog(LOG,
-			 "deleted " UINT64_FORMAT " row(s) from materialization table \"%s.%s\"",
-			 SPI_processed,
-			 NameStr(*materialization_table.schema),
-			 NameStr(*materialization_table.name));
+static void
+emit_materialization_delete_error(MaterializationContext *context)
+{
+	elog(ERROR,
+		 "could not delete old values from materialization table \"%s.%s\"",
+		 NameStr(*context->materialization_table.schema),
+		 NameStr(*context->materialization_table.name));
+}
 
-	rows_processed += SPI_processed;
+static void
+emit_materialization_merge_error(MaterializationContext *context)
+{
+	elog(ERROR,
+		 "could not merge old values into materialization table \"%s.%s\"",
+		 NameStr(*context->materialization_table.schema),
+		 NameStr(*context->materialization_table.name));
+}
 
-	return rows_processed;
+static void
+emit_materialization_insert_progress(MaterializationContext *context, uint64 rows_processed)
+{
+	elog(LOG,
+		 "inserted " UINT64_FORMAT " row(s) into materialization table \"%s.%s\"",
+		 rows_processed,
+		 NameStr(*context->materialization_table.schema),
+		 NameStr(*context->materialization_table.name));
+}
+static void
+emit_materialization_delete_progress(MaterializationContext *context, uint64 rows_processed)
+{
+	elog(LOG,
+		 "deleted " UINT64_FORMAT " row(s) from materialization table \"%s.%s\"",
+		 rows_processed,
+		 NameStr(*context->materialization_table.schema),
+		 NameStr(*context->materialization_table.name));
+}
+static void
+emit_materialization_merge_progress(MaterializationContext *context, uint64 rows_processed)
+{
+	elog(LOG,
+		 "merged " UINT64_FORMAT " row(s) into materialization table \"%s.%s\"",
+		 rows_processed,
+		 NameStr(*context->materialization_table.schema),
+		 NameStr(*context->materialization_table.name));
 }
 
-static uint64
-spi_delete_materializations(SchemaAndName materialization_table, const NameData *time_column_name,
-							TimeRange materialization_range, const char *const chunk_condition)
+static MaterializationPlan *
+get_materialization_plan(MaterializationContext *context, MaterializationPlanType plan_type)
+{
+	Assert(plan_type >= PLAN_TYPE_INSERT);
+	Assert(plan_type < _MAX_MATERIALIZATION_PLAN_TYPES);
+
+	MaterializationPlan *materialization = &materialization_plans[plan_type];
+
+	if (materialization->plan == NULL)
+	{
+		char *query = materialization->create_statement(context);
```

Comment:
Since this is allocated in the memory context, it is not strictly needed to free it after you're done, but it is a good practice to to do that. Since the query will just be allocated once for each plan, it should not be an issue.

Reply:
Added the proper pfree